### PR TITLE
Send federation events concurrently

### PR DESCRIPTION
**Pull request by erikjohnston**
This also cherry-picks `Handle exceptions in get_hosts_for_room...` from matrix-org-hotfixes

### CODE REVIEW

File: synapse/federation/transaction_queue.py [hunk 1]
```diff
@@ -207,6 +213,19 @@ def _process_event_queue_loop(self):
 
                     self._send_pdu(event, destinations)
 
+                def handle_room_events(events):
+                    for event in events:
+                        return handle_event(event)
```
Comment (krombel) [inline synapse/federation/transaction_queue.py]: That will handle the first event of events and skip the other events of the queue - even if that is valid or not

Comment (erikjohnston) [reply synapse/federation/transaction_queue.py]: Good catch

File: synapse/federation/transaction_queue.py [hunk 2]
```diff
@@ -169,32 +169,38 @@ def _process_event_queue_loop(self):
...
+                        # Otherwise if the last member on a server in a room is
+                        # banned then it won't receive the event because it won't
+                        # be in the room after the ban.
+                        destinations = yield self.state.get_current_hosts_in_room(
+                            event.room_id, latest_event_ids=[
+                                prev_id for prev_id, _ in event.prev_events
+                            ],
+                        )
+                    except Exception:
+                        logger.exception("Failed to calculate hosts in room")
```
Comment (richvdh) [inline synapse/federation/transaction_queue.py]: why would this happen?

if we're going to do this, surely we should at least log the event id?

File: synapse/federation/transaction_queue.py [hunk 3]
```diff
@@ -207,6 +213,20 @@ def _process_event_queue_loop(self):
...
+                def handle_room_events(events):
+                    for event in events:
+                        yield handle_event(event)
+
+                events_by_room = {}
+                for event in events:
+                    events_by_room.setdefault(event.room_id, []).append(event)
+
+                yield logcontext.make_deferred_yieldable(defer.gatherResults(
+                    [handle_room_events(evs) for evs in events_by_room.itervalues()],
```
Comment (richvdh) [inline synapse/federation/transaction_queue.py]: this looks like it needs a `preserve_fn`

File: synapse/federation/transaction_queue.py [hunk 4]
```diff
@@ -226,7 +226,10 @@ def handle_room_events(events):
                     events_by_room.setdefault(event.room_id, []).append(event)
 
                 yield logcontext.make_deferred_yieldable(defer.gatherResults(
-                    [handle_room_events(evs) for evs in events_by_room.itervalues()],
+                    [
+                        logcontext.preserve_fn(handle_room_events)(evs)
```
Comment (richvdh) [inline synapse/federation/transaction_queue.py]: let's not `preserve_fn` for every event.. Either preserve_fn once, or use `run_in_background` (?)

File: synapse/federation/transaction_queue.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/federation/transaction_queue.py]: It shouldn't happen really, but it empirically seems to have on matrix.org in the past so is probably worth keeping for paranoia's sake.

File: synapse/federation/transaction_queue.py (hunk 4 above)
Comment (erikjohnston) [reply synapse/federation/transaction_queue.py]: So do `logcontext.run_in_background(handle_room_events(evs))`? Isn't that equivalent to `preserve_fn`?

Comment (richvdh) [reply synapse/federation/transaction_queue.py]:
```
logcontext.run_in_background(handle_room_events, evs)
```

And yes it is, except that it doesn't create a new function object on each loop.

**Pull request merged**